`lib.callPackageWith`: Use abort again instead of throw and fix evaluation errors caused by it

infinisil opened the pull request:
## Description of changes

In https://github.com/NixOS/nixpkgs/pull/269356 by @amjoseph-nixpkgs (but really the split-off https://github.com/NixOS/nixpkgs/pull/271123 by @Mic92), `lib.callPackageWith` was switched from an `abort` to a `throw`, which can be caught with `builtins.tryEval`. Because the ofborg evaluation gracefully handles evaluation failures that can be caught, this meant that all `lib.callPackageWith` errors weren't being guarded against by CI anymore, making CI succeed when it really shouldn't have.

This especially affected evaluation with `allowAliases = false`, which triggers the `callPackageWith` error when aliases are used.

Affected merged PRs are:
- https://github.com/NixOS/nixpkgs/pull/276112
- https://github.com/NixOS/nixpkgs/pull/272997
- https://github.com/NixOS/nixpkgs/pull/278560
- https://github.com/NixOS/nixpkgs/pull/167670
- https://github.com/NixOS/nixpkgs/pull/267247
- https://github.com/NixOS/nixpkgs/pull/275882

There may be more PRs in the pipeline that would fail on master.

@jtojnar already made a fix for one evaluation failure in https://github.com/NixOS/nixpkgs/pull/278744. This PR fixes the remaining errors and uses `abort` again to prevent against future breakages.

In particular `nix-build pkgs/test/release` continues succeeding with this PR.

Ping @lilyinstarlight

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
abcde: Fix evaluation

Was broken in 9fddd791c6a5aab0386eb789521495d0732dc4b4
but not caught due to https://github.com/NixOS/nixpkgs/pull/271123#discussion_r1442134594

infinisil committed:
pythonPackages.dnf-plugins-core: Fix eval with allowAliases false

This was not caught when the package was introduced because of
https://github.com/NixOS/nixpkgs/pull/271123#discussion_r1442134594

infinisil committed:
mate.mate-user-share: Fix evaluation with allowAliases false

Was broken by feb114f7f0f01295a99c108ed2d97cd0d152f41a
And not caught due to https://github.com/NixOS/nixpkgs/pull/271123#discussion_r1442134594

infinisil committed:
pythonPackages.sqlalchemy_1_4: Fix evaluation with allowAliases false

Was broken since 7dd34690ec816fe50624466e1c59ce32c2a43b8a
but not caught due to https://github.com/NixOS/nixpkgs/pull/271123#discussion_r1442134594

infinisil committed:
tests.checkpoint-build: Fix evaluation with allowAliases false

Was broken since introduced, but not caught due to
https://github.com/NixOS/nixpkgs/pull/271123#discussion_r1442134594

infinisil committed:
lib.callPackageWith: Use abort, not throw

This reverts f8ea911f7c4e44b167d4b1b51f6d00ebd93e1ed1, see also https://github.com/NixOS/nixpkgs/pull/271123#discussion_r1442134594

infinisil committed:
cudaPackages.autoAddCudaCompatRunpathHook: Fix eval w/o cuda_compat

yannham commented on pkgs/development/cuda-modules/setup-hooks/extension.nix:
```diff
@@ -53,7 +53,7 @@ final: _: {
     final.callPackage
       (
-        {makeSetupHook, cuda_compat}:
+        {makeSetupHook, cuda_compat ? throw "autoAddCudaCompatRunpathHook: No cuda_compat for CUDA ${final.cudaMajorMinorVersion}" }:
```
So to make sure I understand correctly, it's ok to use `throw` here because we do want to prevent Ofborg from evaluating the hook on platforms that don't support `cuda_compat`? Or should this be an abort as well?

infinisil replied:
Indeed that's right :)

yannham reviewed (approved):
LGTM on the CUDA part. Thanks!

infinisil committed:
pythonPackages.snakemake-interface-common: Fix eval with allowAliases false

Was broken since initialised in 9b2d4d2faa875d91f76d77a9193ff7a840f263e7
but not noticed due to https://github.com/NixOS/nixpkgs/pull/271123#discussion_r1442134594

infinisil commented:
I just rebased and pushed another commit to fix another failure that's since been introduced on master. I'll merge this PR as soon as ofborg passes so we fix this for PRs going forward.

Note that even after merging, new failures on master may be introduced from PRs that already passed CI.

vcunat commented:
This merge itself caused breakage and blocks channels now:
https://hydra.nixos.org/build/245871962

infinisil commented:
@vcunat Yeah this is what I was anticipating with my previous message:

> Note that even after merging, new failures on master may be introduced from PRs that already passed CI.

In the future we should apply the [ratchet check](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test/nixpkgs-check-by-name/README.md#ratchet-checks) idea I recently introduced for `pkgs/by-name` more widely. This would fix such problems.

For now we can only fix forward as these problems happen, I'll do so.

infinisil commented:
Oh actually no this is a different problem, I think this is something that wasn't anticipated in https://github.com/NixOS/nixpkgs/pull/269356, I'll investigate..

vcunat commented:
To be clear, this is broken on the merge commit itself already, not by any subsequent merges.

infinisil commented:
I tracked down the root problem to https://github.com/NixOS/nixpkgs/pull/276800.

Fixed with https://github.com/NixOS/nixpkgs/pull/279684
